File: 1020-number-of-enclaves/1020-number-of-enclaves.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& grid, int i, int j) {
        if (i < 0 || j < 0 || i == grid.size() || j == grid[i].size() || grid[i][j] != 1)
            return;
        grid[i][j] = 0;
        dfs(grid, i + 1, j);
        dfs(grid, i - 1, j);
        dfs(grid, i, j + 1);
        dfs(grid, i, j - 1);
    }
    
    int numEnclaves(vector<vector<int>>& grid) {
        for (auto i = 0; i < grid.size(); i++) {
            for (auto j = 0; j < grid[0].size(); j++) {
                if (i * j == 0 || i == grid.size() - 1 || j == grid[i].size() - 1)
                    dfs(grid, i, j);
            }
        }
        return accumulate(begin(grid), end(grid), 0, [](int s, vector<int>& r) {
            return s + accumulate(begin(r), end(r), 0); });
        }
};

File: 1020-number-of-enclaves/1020-number-of-enclaves.java
class Solution {
    public int numEnclaves(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == 1 && (i == 0 || i == rows - 1 || j == 0 || j == cols - 1)) {
                    dfs(i, j, grid);
                }
            }
        }
        int count = 0;
        for (int[] r : grid) {
            for (int c : r) {
                if (c == 1) {
                    count++;
                }
            }
        }
        return count;
    }

    private void dfs(int r, int c, int[][] grid) {
        if (r < 0 || r >= grid.length || c < 0 || c >= grid[0].length || grid[r][c] != 1) {
            return;
        }
        grid[r][c] = 0;

        dfs(r + 1, c, grid);
        dfs(r - 1, c, grid);
        dfs(r, c + 1, grid);
        dfs(r, c - 1, grid);
    }
}